README
This is an expanded version of the Python debugger, pdb.py

Largely we're trying to follow gdb's command set unless there's good
reason. So new or modified command names generally refer to the
function they have in gdb.

Changes from pdb.py:

- non-interactive POSIX-shell like line tracing.

- Restart program saving settings (gdb's "run") or a pure "exec"
  restart (reloads the entire debugger)

- Add "display" expressions

- Some gdb "set"/"show" commands:
   args, listsize, version (no set here), and some new ones
  basename, cmdtrace, interactive, linetrace, 

- Some gdb "info" commands:
   args break, display, globals, line, locals, program, and source

- Other gdb commands: cd, display, disassemble, frame, pwd, return
  (immediate return, not finish), run, shell, source, undisplay

- debugger is installed in binary directory; i.e. one runs "pydb script..."
  rather than "python pdb script..."

- debugger accepts debugger options; e.g. redirecting output or setting 
  line tracing

- Add Perl's "examine" to show info about a symbol.  For functions,
  methods, classes and modules the documentation string if any is printed.
  For functions, we also show the argument list. More work is needed here
  to recurse fields of an arbitrary object and show that.

- add gdb's "frame" command and stack order. Cleaner display of
  call stack; for example "exec" calls are marked as such.

- gdb-like help with subcomamnd-specific help available on 
  info, set, or show subcommands;  e.g. "help info line" works

- More extensive and complete documentation. See
  http://bashdb.sourceforge.net/pydb/pydb/lib/index.html

- up/down can take a number of frames to move.

- Stepping skips over "def" statements.

- Comes with regression tests. Can also be subclassed to capture
  debugger output.

- Works with GUI frontend ddd. (For now a testing version of ddd is
  required. See
  http://sourceforge.net/project/showfiles.php?group_id=61395&package_id=65341
  version 3.3.12-test3 or greater.

- Miscellaneous pdb bugs fixed. (See ChangeLog for details)

Coexistence: 
  For now, we try not to conflict with pdb.py. After all, Python 
  developers made provision of multiple debuggers so we'll make use of
  that!

  So whereever there was a pdb, use pydb. The class name is still
  Pdb. So the debugger initialization script is .pydbrc for this
  debugger rather than .pdbrc for the one that comes with Python.

$Id: README,v 1.6 2006/02/21 21:12:57 rockyb Exp $

NEWS
1.27git

* Status event for breakpoints is xx
* Remove message that you can step or continue through a post-mortem to restart
* Fix off-by-one bug in finish
* Add an "edit" command.

1.26
04-10-2009
* Add -d (debug debugger) option to ipython
* Add pdef command
* Allow a signal number in "info signals"
* Disassemble command improvements, such as disassembly 
  of frames
* Add "p" format specifier in "print" command
* Add a skip (statement) command

The above are all backports from pydbgr. 

* Doc fixes

1.25
12-25-2008
* Clean up debugger(). Add "status=continue".  
* list now works in the presense of zipped eggs
* Some small bug fixes and configuration bug fixes

1.24
12-10-2008
* "frame" (no arg) is "frame 0" like gdb

* Fix a bug starting ipython and running pydb.

* Catch exceptions raised in runcall()

* Revise tests for 2.6 and systems without readline

* Small doc and test changes. 

This may be the last release before a major rewrite.

1.23
5-30-2008

* Show parameters on call

* Some doc fixes.

* add ipython and python commands to go into an ipython or a python shell

* add "save" command to save the current breakpoints and/or 
  most settings to a file. (Suggested at a NYC Python group meeting.)

* add set/show "deftrace" to allow stopping at a def (method creation)
  statement.

* pseudo gdb-like "annotate" set/show commands and option.

* Better emacs interaction via above annotate

* bug when "info local" is run inside a "with" statement

  Fixes for last three items based patches by Alberto Griggio

1.22 2007-04-15

* Add function tracing and option -F.

* Call and return trace lines show nesting level. Return will show the
  value if it is a scalar or string type or return type otherwise.

* Make pydb.el work on Emacs 22 and 23. (was previously just 21).
  Better emacs tolerance on MS Windows.

* fix bug in "info thread" when threadframe is not installed (and
  python < 2.5).

* Better ipython interaction.

* pm distinquishes now where an exception rolled back to versus where
  the exception occurred.

* BSD Make tolerance

1.21 2007-02-13

* pydb script will do a path search to find a script, same as gdb.
  If compiled Python script is given, we try to find the non-compiled
  equivalent.

* 'help' command now allows an expression and runs Python's builtin help
  command (pydoc). Add pydoc command to call pydoc externally.

* Make prompt string detection in Emacs pydb-track more MS-friendly.
  Add a unit test for the Emacs interface.

* Allow lowercase signal names, e.g. "int" or "sigint" for "SIGINT".

* disassemble command allows for a start line and end line to limit
  disassembly, same as gdb. Due to these changes, we now require at least 
  Python 2.4.0.

* Allow limited expressions many places an int parameter is used in
  a debugger command

* Allow "info thread" always. That is we don't need --threading for this

* signal command's output more like gdb's.

* Add pydoc's help (i.e. python's built-in help) to debugger "help"
  command

* Add "set"/"show" option to force debugger output to be flushed on
  write.

* set/show "cmdtrace" is now "trace-commands" to match gdb 6.6.

* work-around a possible Python 2.5 bug manifested in pm/post_mortem where
  t.tb_frame.f_lineno != t.tb_lineno

* Experimental command completion. "complete" command now adds in local and
  global names on the first word

* Handle a quit when called via pydb.debugger(). Also experimental.

* We now print the line text and (and again the line number) when we stop.
  This is more helpful and more consistent with gdb and other debuggers.
  Note linetrace adds also (redundant) line number.

1.20 2006-12-10

* Turning on signal checking may slow things down too much. So off
  by default. Fix bug in "set sigcheck". Add --sigcheck option.

* Allow a thread number in "info thread" and "frame" commands if 
  Python 2.5 or threadframe.

* Show all threads even if we don't know their names.

* Allow Pdb object to get passed in runv(). Allows for better ipython
  invocation (%pydb) and may be useful in other shells as well

* Provide a clue when debugger "run" command given but no "file" has
  been set.

* Better shell-like argument parsing in the "run" command via
  shlex. 

* Recover more gracefully if a post-mortem is called and there is none.

* "set_trace" is now an alias for "debugger" which is probably a more
  natural name.

* We now tolerate back to Python version 2.3.3

* A number of small bugs fixed and document updates.

1.19 2006-10-27

* info threads: merge in terse display (Python 2.5, threadframe)
  and sort by thread name (Python 2.5)

* Add optional position argument on "continue" command.

* Add pydb.runl() and pydb.runv() forms to facilitate debugging a
  program from inside a Python shell.

* fix thread bugs: on startup with threading we stopped one statement
  too soon. more careful about storing into variables used in
  threading which are not locked (and so could be changed during a
  thread switch)

* Allow quit in thread debugging when MainThread is the only thread
  that we are in

* fix bugs in signal handling. 

* Add "stack" and "nostack" to list of things that can be done when
  a signal is received.

* things to make more ipython friendly.

* Improve pydb manual

1.18 2006-09-25

* Start thread debugging: 
    info thread [thread-name] [terse|verbose]    - show thread info
    qt                                     - quit current thread
    can set a breakpoint on a specific thread name.
  expand definitions of "frame" and "break" for threading

* GDB-like signal handling. Some code based on Matt Fleming's 2006
  Summer-of-Code project. Gdb's "handle" and "signal" commands.

* Add hard kill. May be needed for thread debugging.

* Fix bug in handling temporary breakpoints (tbreak) caused by the
  difference in semantics between pdb's 'clear' command and pdb's
  (which calls this 'delete' and has a different meaning for 'clear').


pydb/subcmd.py
"""Handles gdb-like subcommand processing."""
__revision="$Id: subcmd.py,v 1.9 2007/02/04 13:00:12 rockyb Exp $"
# -*- coding: utf-8 -*-
#   Copyright (C) 2006, 2007 Rocky Bernstein
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
#    02110-1301 USA.

class Subcmd:
    """Gdb-like subcommand handling """
    def __init__(self, name, doc):
        self.name    = name
        self.doc     = doc
        self.subcmds = {}
        self.cmdlist = []

    def lookup(self, subcmd_prefix):
        """Find subcmd in self.subcmds"""
        for subcmd_name in self.subcmds.keys():
            if subcmd_name.startswith(subcmd_prefix) \
               and len(subcmd_prefix) >= self.subcmds[subcmd_name]['min']:
                return self.subcmds[subcmd_name]
        return None

    def _subcmd_helper(self, subcmd_name, obj, label=False, strip=False):
        """Show help for a single subcommand"""
        if label:
            obj.msg_nocr("%s %s --" % (self.name, subcmd_name))

        entry = self.lookup(subcmd_name)
        if entry:
            d = entry['doc']
            if strip:
                # Limit the help message to one line (delimited by '\n')
                if '\n' in d:
                    d = d[:d.find('\n')]
                # If the last character is a period, remove it.
                if d[-1] == '.': d = d[:d.find('.')]
            obj.msg(d)
            return
        obj.undefined_cmd("help", subcmd_name)

    def add(self, subcmd_name, subcmd_cb, min_len=0, in_list=True):
        """Add subcmd to the available subcommands for this object.
        It will have the supplied docstring, and subcmd_cb will be called
        when we want to run the command. min_len is the minimum length
        allowed to abbreviate the command. in_list indicates with the
        show command will be run when giving a list of all sub commands
        of this object. Some commands have long output like "show commands"
        so we might not want to show that.
        """
        self.subcmds[subcmd_name] = {
            "callback": subcmd_cb,
            "name"    : subcmd_name,
            "doc"     : subcmd_cb.__doc__,
            "in_list" : in_list,
            "min"     : min_len}

        # We keep a list of subcommands to assist command completion
        self.cmdlist.append(subcmd_name)

    def do(self, obj, subcmd_name, arg):
        """Run subcmd_name with args using obj for the environent"""
        entry=self.lookup(subcmd_name)
        if entry:
            entry['callback'](arg)
        else:
            obj.undefined_cmd(self.name, subcmd_name)

    # Note: format of help is compatible with ddd.
    def help(self, obj, *args):
        """help for subcommands."""

        subcmd_prefix = args[0]
        if not subcmd_prefix or len(subcmd_prefix) == 0:
            obj.msg(self.doc)
            obj.msg("""
List of %s subcommands:
""" % (self.name))
            for subcmd_name in self.list():
                self._subcmd_helper(subcmd_name, obj, True, True)
            return

        entry = self.lookup(subcmd_prefix)
        if entry:
            self._subcmd_helper(entry['name'], obj)
        else:
            obj.errmsg("Unknown 'help %s' subcommand %s"
                       % (self.name, subcmd_prefix))

    def list(self):
        l = self.subcmds.keys()
        l.sort()
        return l


# When invoked as main program, invoke the debugger on a script
if __name__ == '__main__':

    class FakeGdb:
        "A Mock Gdb class"
        def msg_nocr(self, msg): print msg,
        def msg(self, msg): print msg
        def errmsg(self, msg): print msg

        def info_args(self, arg):
            "Print the arguments of the current function."
            print "a=1, b=2"
        def info_break(self, arg):
            "Without argument, list info about all breakpoints"
            print "no breakpoints"
        def set_args(self, arg):
            """Set argument list to give program being debugged when it is started"""
            print "Not done yet"
        def set_basename(self, arg):
            "Set short filenames (the basename) in debug output"
            print "basename changed to %s" % arg

        def show_args(self, arg):
            """Show argument list to give debugged program on start"""
            print "Argument list to give program is ''"
        def show_basename(self, arg):
            "Show if we are to show short of long filenames"
            print "basename is off."
        def show_cmdtrace(self, arg):
            "Show if we are to show debugger commands"
            print "cmdtraces is on."


    gdb = FakeGdb()
    infocmd = Subcmd('info',
"""Generic command for showing things about the program being debugged. """)

    infocmd.add('args', gdb.info_args)
    infocmd.add('break', gdb.info_break)
    showcmd = Subcmd('show',
    """Generic command for showing things about the debugger.""")
    showcmd.add('args', gdb.show_args)
    showcmd.add('basename', gdb.show_basename,)
    showcmd.add('cmdtrace', gdb.show_cmdtrace)

    showcmd.help(gdb, '')
    print "-" * 20
    showcmd.help(gdb, "args")
    print "-" * 20
    showcmd.do(gdb, "args", "")
    print "-" * 20
    infocmd.help(gdb, '')
    print "-" * 20
    infocmd.help(gdb, 'basename')
    setcmd = Subcmd('set',
    """This command modifies parts of the debugger environment.
You can see these environment settings with the 'show' command.""")

    setcmd.add('args', gdb.set_args)
    setcmd.add('basename', gdb.set_basename)
    print "-" * 20
    setcmd.help(gdb, '')
    print "-" * 20
    setcmd.help(gdb, 'basename')
    print "-" * 20
    setcmd.do(gdb, 'basename', 'off')

    print showcmd.list()

